### PR TITLE
Recalculate duties when a reorg occurs

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Rework `DutyScheduler` to better separate concerns.
When a reorg event occurs, discard all scheduled duties and recompute them.

### CODE REVIEW

File: validator/client/src/main/java/tech/pegasys/artemis/validator/client/ValidatorClientService.java [hunk 1]
```diff
@@ -48,10 +48,14 @@ public static ValidatorClientService create(final ServiceConfig config) {
         ValidatorLoader.initializeValidators(config.getConfig());
     final ValidatorDutyFactory validatorDutyFactory =
         new ValidatorDutyFactory(forkProvider, validatorApiChannel);
-    final ScheduledDuties scheduledDuties = new ScheduledDuties(validatorDutyFactory);
-    final DutyScheduler validatorClient =
+    final ValidatorTimingChannel validatorClient =
```
Comment (mbaxter) [inline validator/client/src/main/java/tech/pegasys/artemis/validator/client/ValidatorClientService.java]:
```suggestion
    final ValidatorTimingChannel validatorTimingChannel =
```

File: validator/client/src/main/java/tech/pegasys/artemis/validator/client/EpochDutiesScheduler.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import tech.pegasys.artemis.validator.api.NodeSyncingException;
+import tech.pegasys.artemis.validator.api.ValidatorApiChannel;
+import tech.pegasys.artemis.validator.api.ValidatorDuties;
+import tech.pegasys.artemis.validator.client.duties.ScheduledDuties;
+
+class EpochDutiesScheduler {
+  private static final Logger LOG = LogManager.getLogger();
+  private final AsyncRunner asyncRunner;
+  private final ValidatorApiChannel validatorApiChannel;
+  private final ForkProvider forkProvider;
```
Comment (mbaxter) [inline validator/client/src/main/java/tech/pegasys/artemis/validator/client/EpochDutiesScheduler.java:39]: Do we need to make sure to refresh the fork info when a reorg occurs?

File: validator/client/src/main/java/tech/pegasys/artemis/validator/client/DutyScheduler.java [hunk 3]
```diff
@@ -16,232 +16,65 @@
...
-import tech.pegasys.artemis.validator.api.NodeSyncingException;
-import tech.pegasys.artemis.validator.api.ValidatorApiChannel;
-import tech.pegasys.artemis.validator.api.ValidatorDuties;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.function.BiConsumer;
 import tech.pegasys.artemis.validator.api.ValidatorTimingChannel;
-import tech.pegasys.artemis.validator.client.duties.ScheduledDuties;
 
 public class DutyScheduler implements ValidatorTimingChannel {
```
Comment (mbaxter) [inline validator/client/src/main/java/tech/pegasys/artemis/validator/client/DutyScheduler.java:25]: Nice refactor :D

**Review by mbaxter — APPROVED** [review]
LGTM

File: validator/client/src/main/java/tech/pegasys/artemis/validator/client/EpochDutiesScheduler.java (hunk 2 above)
Comment (ajsutton) [reply validator/client/src/main/java/tech/pegasys/artemis/validator/client/EpochDutiesScheduler.java:39]: I don't think so.  The Fork data includes both "current" and "previous" forks and which epoch "current" takes effect.  Forks would be scheduled well in advance (and would require upgraded client software) so even if the reorg activated or deactivated a fork, the actual `Fork` data would still be the same - just a question of whether we are now before or after the appropriate epoch.

**Pull request merged**